Add Meilisearch example using a CLI (#49)
thedataquarry/rustinpieces: pieces/meilisearch_with_cli/README.md

# Interact with Meilisearch though a CLI

Update settings, index data, and search using [Meilisearch](https://www.meilisearch.com/) with a CLI.

## Goal

Meilisearch is a fast, typo-tolerate search engine with a focus on being easy to setup and use. In
Meilisearch, indexes are created to store documents and their associated settings. These indexes are
simlilar to tables in relational databases or collections in document databases.

In this example we will create and update the settings for an index, send documents to the server for
indexing, and preform searaches all through a CLI.

For data we will use the [wine reviews dataset from Kaggle](https://www.kaggle.com/datasets/zynicide/wine-reviews).
The dataset is in the following format:

> ```json

Reviewer: ```suggestion
```

Reviewer: There's a rendering error because `null` isn't valid JSON for the syntax highlighter. I recommend leaving the `json` part out.

# Interact with Meilisearch though a CLI

Update settings, index data, and search using [Meilisearch](https://www.meilisearch.com/) with a CLI.

## Goal

> Meilisearch is a fast, typo-tolerate search engine with a focus on being easy to setup and use. In

Reviewer: ```suggestion
[Meilisearch](https://github.com/meilisearch/meilisearch) is a fast, open source, typo-tolerant search engine with a focus on being easy to setup and use. In
```

Meilisearch, indexes are created to store documents and their associated settings. These indexes are

> simlilar to tables in relational databases or collections in document databases.

Reviewer: ```suggestion
similar to tables in relational databases or collections in document databases.
```

In this example we will create and update the settings for an index, send documents to the server for

> indexing, and preform searaches all through a CLI.

Reviewer: ```suggestion
indexing, and perform searches all through a CLI.
```

For data we will use the [wine reviews dataset from Kaggle](https://www.kaggle.com/datasets/zynicide/wine-reviews).
The dataset is in the following format:

> ```

Reviewer: Missing the opening curly brace here.

  id: number | null,
  title: string,
  description: string,
  taster_name: string | null,
  designation: string | null,
  variety: string | null,
  country: string | null,
  winery: string | null,
  points: string | null,
  taster_twitter_handle: string | null,
  price: number | null,
  region_1: string | null,
  region_2: string | null,
  province: string | null,
}
```

We will update the searchable attributes to include `title`, `description`, `taster_name`, `designation`,
`variety`, `province`, `country`, and `winery`. Reducing the number of searchable fields will make indexing
faster because we don't need to waste time indexing fields that aren't going to be searched.

Additionally we will make `title` and `country` sortable files. The results in Meilisearch are ordered
based on ranking rules with the default rules being: `["words", "typo", "proximity", "attribute", "sort", "exactness"]`.
The result of this is records are only sorted if `words`, `typo`, `proximity`, and `attribute` all

> match for multiple records. Because we want sort to take precidence we will update the ranking rules

Reviewer: ```suggestion
match for multiple records. Because we want sort to take precedence we will update the ranking rules
```

│ --limit       -l      INTEGER  Limit the number of search results [default: 20]                                                                                                           │
│ --sort        -s      TEXT     Sort order for the results [default: None]                                                                                                                 │
│ --index-name  -i      TEXT     The name to use for the index [default: wine]                                                                                                              │
│ --help                         Show this message and exit.                                                                                                                                │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

### Takeaways

Typically in a CLI like this asyncio will not provide any benefits, and may actually be slower

> because of the need to create an event loop. An exception to that in this case is `meilisearch-python-sdk`

Reviewer: ```suggestion
because of the need to create an event loop. An exception in this case is that `meilisearch-python-sdk`
```

> has an optomization that sends documents for indexing concurrently when adding them in batches,

Reviewer: ```suggestion
has an optimization that sends documents for indexing concurrently when adding them in batches,
```

as we are here, if asyncio is used. Because of this we take advantage of `meilisearch-python-sdk`
providing both and `AsyncClient` and a `Client` to use asyncio when indexing the documents, while

> preforming the index creation and searches syncronously.

Reviewer: ```suggestion
performing the index creation and searches synchronously.
```

  -h, --help                     Print help

$ cargo run -- index-data --help
Index the wine data

Usage: meilisearch-cli index-data [OPTIONS]

Options:
  -d, --data-path <DATA_PATH>    Index the wine data
  -i, --index-name <INDEX_NAME>  The name to use for the index. Default: wine
  -w, --wait                     Wait for the data to finish indexing
  -h, --help                     Print help

$ cargo run -- search --help
Preform a search

Usage: meilisearch-cli search [OPTIONS] <QUERY>

Arguments:
  <QUERY>  The search to preform

Options:
  -l, --limit <LIMIT>            Limit the number of search results
  -s, --sort <SORT>              Sort order for the results
  -i, --index-name <INDEX_NAME>  The name to use for the index. Default: wine
  -h, --help                     Print help
```

### Takeaways

In Rust asyncio is run using crates, of which you have options to chose from. `meilisearch-rust` was
designed to let the user pick which async crate to use rather than dictating one. In this example we
use async-std, but we could have just as easily used tokio or some other crate. It is possilbe to make

> the program syncronous `meilisearch-rust` using the [futures](https://github.com/rust-lang/futures-rs)

Reviewer: ```suggestion
the program synchronous in `meilisearch-rust` using the [futures](https://github.com/rust-lang/futures-rs)
```

> crate `block_on`, however it is not as simple to mix asyncio and syncronous calls as we have done in

Reviewer: ```suggestion
crate `block_on`, however it is not as simple to mix asyncio and synchronous calls as we have done in
```

the Python example. Because of this we choose to make everything async in the Rust examples.

As with the Python example, we create indexes using uuids for the names in the tests. Unlike in Python
where the tests are run syncronously, the Rust test runner runs tests in parallel. This makes the
use of uuids here more important in Rust. Without using unique names for the indexes we would end up
with race conditions in the tests. Sometimes they would pass and sometimes they would fail without an
obvious reason for why. This is because one test could be using the index from another test by chance.

## Performance

One big performance advantage Rust has over Python in CLIs, as a compiled language, is startup time.

> With Python there is overhead associated with starting the Python interperter which causes a noticable

Reviewer: ```suggestion
With Python, there is overhead associated with starting the Python interpreter which causes a noticeable
```

delay in startup time, while Rust is almost instant. An informal test using the `time` command illustrates
this well. For `create-index`, the Rust program runs in an average of 184ms while the Python program
average is 796ms on a System76 Lemur Pro with a 13th Gen Intel Core i7-1355U processor and 40GB of
RAM.

To further illustrate this, for `index-data` the average run time for Rust is 3.41s of which 3.29s is
spent doing the actual indexing. The average run time for Python is 3.03s of which 1.19s is spent on
the indexing. Because of the async batch indexing optimizations in `meilisearch-python-async` the Python

> example is significatnly faster than the Rust example at indexing, however almost all of this advantage

Reviewer: ```suggestion
example is significantly faster than the Rust example at indexing, however almost all of this advantage
```
